File: app/src/main/java/com/example/mike/mp3player/client/views/ThemeSpinnerController.java
package com.example.mike.mp3player.client.views;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.res.Resources;
import android.content.res.TypedArray;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Spinner;

import androidx.annotation.NonNull;
import androidx.annotation.VisibleForTesting;

import com.example.mike.mp3player.R;
import com.example.mike.mp3player.client.activities.MainActivityInjector;
import com.google.common.collect.BiMap;
import com.google.common.collect.HashBiMap;

import java.util.ArrayList;
import java.util.List;

import static android.content.Context.MODE_PRIVATE;
import static com.example.mike.mp3player.commons.Constants.THEME;

public class ThemeSpinnerController implements AdapterView.OnItemSelectedListener {

    private static final int[] attrs = {R.attr.themeName};
    private static final String LOG_TAG = "THM_SPNR_CTLR";
    private final Activity activity;
    private Context context;
    private Spinner spinner;
    private ArrayAdapter<String> adapter; // TODO: make a make from Theme name to resource
    private List<Integer> themeResIds;
    private BiMap<String, Integer> themeNameToResMap;
    private long selectCount = 0;
    private String currentTheme;

    public ThemeSpinnerController(@NonNull Context context, @NonNull Spinner spinner, @NonNull Activity activity) {
        this.context = context;
        this.spinner = spinner;
        this.activity = activity;
        init();
    }

    private void init() {
        this.adapter = new ArrayAdapter<>(context, android.R.layout.simple_spinner_dropdown_item);
        spinner.setAdapter(adapter);
        spinner.setOnItemSelectedListener(this);
        TypedArray themeArray = context.getResources().obtainTypedArray(R.array.themes);
        this.themeResIds = new ArrayList<>();
        this.themeNameToResMap = HashBiMap.create();

        if (themeArray != null && themeArray.length() > 0) {
            int numberOfResources = themeArray.length();
            for (int i = 0; i < numberOfResources; i++) { // for each theme in the theme array
                int res = themeArray.getResourceId(i, 0);
                themeResIds.add(res);
                TypedArray themeNameArray = context.obtainStyledAttributes(res, attrs); // get the theme name GIVEN the themes res if.
                String themeName = themeNameArray.getString(0);
                adapter.add(themeName);
                getThemeNameToResMap().put(themeName, res);
                recycleTypedArray(themeNameArray);
            }
        }

        recycleTypedArray(themeArray);
        int currentThemeId = getCurrentThemeId();
        if (currentThemeId != -1) {
            currentTheme = getThemeNameToResMap().inverse().get(currentThemeId);
            int position = adapter.getPosition(getCurrentTheme());
            spinner.setSelection(position);
        }
    }


    @Override
    public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
        int res = themeResIds.get(position);
        Log.d(LOG_TAG, "selected " + getThemeNameToResMap().get(res));
        if (selectCount >= 1) {
            Log.d(LOG_TAG, "select count > 1");
            setThemePreference(res);
            activity.finish();
            Intent intent = new Intent(context, MainActivityInjector.class);
            intent.putExtra(THEME, res);
            activity.startActivity(intent);
        }
        selectCount++;
    }

    @Override
    public void onNothingSelected(AdapterView<?> parent) {
    }

    /**
     *
     * @return
     */
    private int getCurrentThemeId() {
        Resources.Theme activityTheme = activity.getTheme();
        if (null != activityTheme) {

            TypedArray themeNameArray = activityTheme.obtainStyledAttributes(attrs);
            if (null != themeNameArray && themeNameArray.length() > 0) {
                String themeName = themeNameArray.getString(0);
                themeNameArray.recycle();
                Log.d(LOG_TAG, "current theme is: " + themeName);
                Integer result = getThemeNameToResMap().get(themeName);
                return result == null ? -1 : result;
            }
        }
        return  -1;
    }
    /**
     *
     * @param themeId the theme id
     */
    private void setThemePreference(int themeId) {
        SharedPreferences settings = context.getSharedPreferences(THEME, MODE_PRIVATE);
        if (settings != null) {
            SharedPreferences.Editor editor = settings.edit();
            if (editor != null) {
                editor.putInt(THEME, themeId);
                editor.apply();
            }
        }
    }
    /**
     *
     */
    private void recycleTypedArray(TypedArray typedArray) {
        if (null != typedArray) {
            typedArray.recycle();
        }
    }

    @VisibleForTesting
    public BiMap<String, Integer> getThemeNameToResMap() {
        return themeNameToResMap;
    }

    public String getCurrentTheme() {
        return currentTheme;
    }
}

File: client/src/main/java/com/github/goldy1992/mp3player/client/views/ThemeSpinnerController.kt
package com.github.goldy1992.mp3player.client.views

import android.app.Activity
import android.content.Context
import android.content.res.TypedArray
import android.util.Log
import android.view.View
import android.widget.AdapterView
import android.widget.AdapterView.OnItemSelectedListener
import android.widget.ArrayAdapter
import android.widget.Spinner
import com.github.goldy1992.mp3player.client.R
import com.github.goldy1992.mp3player.commons.Constants
import com.github.goldy1992.mp3player.commons.LogTagger
import com.google.common.collect.BiMap
import com.google.common.collect.HashBiMap
import dagger.hilt.android.qualifiers.ActivityContext
import java.util.*
import javax.inject.Inject

class ThemeSpinnerController
    @Inject
    constructor(@ActivityContext private val context: Context)
    : OnItemSelectedListener, LogTagger {

    private var adapter  : ArrayAdapter<String>? = null
    private var themeResIds: MutableList<Int>? = null
    val attrs = intArrayOf(R.attr.themeName)


    var themeNameToResMap: BiMap<String, Int> = HashBiMap.create()

    private var selectCount: Long = 0
    var currentTheme: String? = null
        private set


    private lateinit var spinner: MaterialDropdownList

    fun init(spinner: MaterialDropdownList) {
        this.spinner = spinner
        adapter = ArrayAdapter(context, R.layout.material_dropdown_item)
        spinner.adapter = adapter
        spinner.addOnItemSelectedListener(this)
        val themeArray = context.resources.obtainTypedArray(R.array.themes)
        themeResIds = ArrayList()
        if (themeArray.length() > 0) {
            val numberOfThemes = themeArray.length()
            for (i in 0 until numberOfThemes) { // for each theme in the theme array
                val res = themeArray.getResourceId(i, 0)
                themeResIds?.add(res)
                val themeNameArray = context.obtainStyledAttributes(res, attrs) // get the theme name GIVEN the themes res if.
                val themeName = themeNameArray.getString(0)
                adapter!!.add(themeName)
                themeNameToResMap[themeName] = res
                recycleTypedArray(themeNameArray)
            }
        }
        recycleTypedArray(themeArray)
        val currentThemeId = currentThemeId
        if (currentThemeId != -1) {
            currentTheme = themeNameToResMap.inverse()[currentThemeId]
            val position = adapter!!.getPosition(currentTheme)
         //   spinner.setSelection(position)
        }

    }

    override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
        val res = themeResIds!![position]
        Log.d(logTag(), "selected " + themeNameToResMap.inverse()[res])
        if (selectCount >= 1) {
            Log.d(logTag(), "select count > 1")
            setThemePreference(res)
            (context as Activity).recreate()
        }
        selectCount++
    }

    override fun onNothingSelected(parent: AdapterView<*>?) = Unit
    /**
     *
     * @return
     */
    private val currentThemeId: Int
        get() {
            val activityTheme = context.theme
            if (null != activityTheme) {
                val themeNameArray = activityTheme.obtainStyledAttributes(attrs)
                if (null != themeNameArray && themeNameArray.length() > 0) {
                    val themeName = themeNameArray.getString(0)
                    themeNameArray.recycle()
                    Log.d(logTag(), "current theme is: $themeName")
                    val result = themeNameToResMap[themeName]
                    return result ?: -1
                }
            }
            return -1
        }

    /**
     *
     * @param themeId the theme id
     */
    private fun setThemePreference(themeId: Int) {
        val settings = context.getSharedPreferences(Constants.THEME, Context.MODE_PRIVATE)
        if (settings != null) {
            val editor = settings.edit()
            if (editor != null) {
                editor.putInt(Constants.THEME, themeId)
                editor.apply()
            }
        }
    }

    /**
     *
     */
    private fun recycleTypedArray(typedArray: TypedArray?) {
        typedArray?.recycle()
    }

    override fun logTag(): String {
       return "THM_SPNR_CTLR"
    }
}